### PR TITLE
Collector

### PR DESCRIPTION
**Pull request by dabonnie**
* Add missing moving_average.cpp file
* Define collector class
* Add relevant unit tests

### CODE REVIEW

File: system_metrics_collector/test/moving_average_statistics/test_moving_average_statistics.cpp [hunk 1]
```diff
@@ -151,21 +151,21 @@ TEST_F(MovingAverageStatisticsTestFixture, test_thread_safe) {
 
   std::thread t1([this, &count, &total_sum]() {
       for (int i = 1; i < 1101.0; i++) {
```
Comment (zmichaels11) [inline system_metrics_collector/test/moving_average_statistics/test_moving_average_statistics.cpp]:
```suggestion
      for (int i = 1; i < 1101; i++) {
```

File: system_metrics_collector/test/moving_average_statistics/test_moving_average_statistics.cpp [hunk 2]
```diff
@@ -151,21 +151,21 @@ TEST_F(MovingAverageStatisticsTestFixture, test_thread_safe) {
...
-        moving_average_statistics->add_measurement(static_cast<double>(i));
+        moving_average_statistics->addMeasurement(static_cast<double>(i));
         count++;
         total_sum += i;
       }
     });
   std::thread t2([this, &count, &total_sum]() {
       for (int i = 1; i < 2101.0; i++) {
```
Comment (zmichaels11) [inline system_metrics_collector/test/moving_average_statistics/test_moving_average_statistics.cpp]:
```suggestion
      for (int i = 1; i < 2101; i++) {
```

File: system_metrics_collector/test/moving_average_statistics/test_moving_average_statistics.cpp [hunk 3]
```diff
@@ -151,21 +151,21 @@ TEST_F(MovingAverageStatisticsTestFixture, test_thread_safe) {
...
-        moving_average_statistics->add_measurement(static_cast<double>(i));
+        moving_average_statistics->addMeasurement(static_cast<double>(i));
         count++;
         total_sum += i;
       }
     });
   std::thread t3([this, &count, &total_sum]() {
       for (int i = 1; i < 3101.0; i++) {
```
Comment (zmichaels11) [inline system_metrics_collector/test/moving_average_statistics/test_moving_average_statistics.cpp]:
```suggestion
      for (int i = 1; i < 3101; i++) {
```

File: system_metrics_collector/CMakeLists.txt [hunk 4]
```diff
@@ -13,8 +13,8 @@ endif()
 
 # Enable strict compiler flags if possible.
 include(CheckCXXCompilerFlag)
-# Removed -Wmissing-declarations until gmock is ignored
-set(FLAGS -pedantic -Wno-long-long -Wall -Wextra -Wcast-qual -Wformat -Wwrite-strings -Wcast-align -Wno-error=cast-align)
+set(FLAGS -pedantic -Wno-long-long -Wall -Wextra -Wcast-qual -Wformat -Wwrite-strings -Wcast-align
+        -Wno-error=cast-align -Wmissing-declarations)
```
Comment (piraka9011) [inline system_metrics_collector/CMakeLists.txt]: Nit in CMake: consistent 2 lines or 8 for indentation

File: system_metrics_collector/src/moving_average_statistics/types.hpp [hunk 5]
```diff
@@ -27,4 +29,19 @@ struct StatisticData
   uint64_t sample_count = 0;
```
[the remark this answers is not in the document]
Comment (dabonnie) [reply system_metrics_collector/src/moving_average_statistics/types.hpp]: Nice catch.

File: system_metrics_collector/src/system_metrics_collector/collector.hpp [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef SYSTEM_METRICS_COLLECTOR__COLLECTOR_HPP_
+#define SYSTEM_METRICS_COLLECTOR__COLLECTOR_HPP_
+
+#include "../../src/moving_average_statistics/moving_average.hpp"
+#include "../../src/moving_average_statistics/types.hpp"
```
[the remark this answers is not in the document]
Comment (dabonnie) [reply system_metrics_collector/src/system_metrics_collector/collector.hpp]: Not quite, that didn't compile. I fixed the header though as it went above one level too much.

File: system_metrics_collector/test/system_metrics_collector/test_collector.cpp [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+    test_collector = std::make_unique<TestCollector>();
+  }
+
+  void TearDown() override
+  {
+    test_collector.reset();
+  }
+
+protected:
+  std::unique_ptr<TestCollector> test_collector = nullptr;
```
Comment (thomas-moulard) [inline system_metrics_collector/test/system_metrics_collector/test_collector.cpp]:
```suggestion
  std::unique_ptr<TestCollector> test_collector{};
```

File: system_metrics_collector/test/moving_average_statistics/test_moving_average_statistics.cpp [hunk 8]
```diff
@@ -181,6 +181,16 @@ TEST_F(MovingAverageStatisticsTestFixture, test_thread_safe) {
...
 }
 
+TEST(MovingAverageStatisticsTest, test_pretty_printing) {
+  StatisticData data;
+  ASSERT_EQ("avg=nan, min=nan, max=nan, std_dev=nan, count=0", statisticsDataToString(data));
+
+  MovingAverageStatistics stats;
+  stats.addMeasurement(1);
+  ASSERT_EQ("avg=1.000000, min=1.000000, max=1.000000, std_dev=0.000000, count=1",
+    statisticsDataToString(stats.getStatistics()));
```
Comment (thomas-moulard) [inline system_metrics_collector/test/moving_average_statistics/test_moving_average_statistics.cpp]: I think this comment was not sent last time somehow - be careful with exact matches, the precision is probably platform dependent, as well as how they are displayed.

**Pull request merged**